components/supervisor/pkg/terminal/terminal.go

```diff
@@ -146,43 +147,56 @@ func (m *Mux) doClose(alias string, gracePeriod time.Duration) error {
 	return nil
 }
 
-func (term *Term) gracefullyShutdownProcess(gracePeriod time.Duration) error {
+func (term *Term) gracefullyShutdownProcess(wg *sync.WaitGroup) error {
 	if term.Command.Process == nil {
-		// process is alrady gone
+		// process is already gone
 		return nil
 	}
-	if gracePeriod == 0 {
-		return term.shutdownProcessImmediately()
-	}
-
-	err := term.Command.Process.Signal(unix.SIGTERM)
-	if err != nil {
-		return err
+	pid := term.Command.Process.Pid
+	var subwg sync.WaitGroup
+	processes, err := procfs.AllProcs()
+	if err == nil {
```

Comment:
I really struggle to parse the code - mainly because of the deep nesting and combination of Go routines and defers.

I'm currently trying to write an equivalent that's a bit more "left aligned". 
Would you be up for a 15min sync session to walk me through?

Reply:
Yes, me too. I've rewritten / left-aligned it. Also added a test.